heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,623 @@
+
+```
+sawscript> print env
+[20:19:48.436] Module: examples/linked_list.bc
+Types:
+  %struct.list64_t = type { i64, %struct.list64_t* }
+
+Globals:
+
+External references:
+  declare default void @llvm.dbg.declare(metadata, metadata,
+                                         metadata)
+  declare default i8* @malloc(i64)
+
+Definitions:
+  i64 @is_head(i64 %0, %struct.list64_t* %1)
+  i64 @is_elem(i64 %0, %struct.list64_t* %1)
+  i64 @any(i64(i64)* %0, %struct.list64_t* %1)
+  %struct.list64_t* @find_elem(i64 %0, %struct.list64_t* %1)
+  %struct.list64_t* @sorted_insert(i64 %0, %struct.list64_t* %1)
+  %struct.list64_t* @sorted_insert_no_malloc(%struct.list64_t* %0,
+                                             %struct.list64_t* %1)
+```
+
+The heapster environment contains all the types, global definitions,
+external references and functions from the loaded module.
+
+
+#### 4. Writing heapster specifications for your functions
+
+##### Pen and paper specification
+
+Before we go foward using Heapster, let's think about what the
+specification for `is_elem` should be. As we can check in the
+environment `env`, `is_elem` takes a 64-bit integer the "element".
+
+```LLVM
+i64 @is_elem(i64 %0, %struct.list64_t* %1)
+```
+
+Hence, precondition to the function should reflect these two inputs
+like so `arg0:int64, arg1: list int64`, for some predicates `int64`
+and `list` that express a single integer and a list of them.
+
+Moreover, the function returns a new i64-bit integer and, as we can
+see in the code of `is_elem`, the arguments are not modified. So the
+postcondition, of the function could be described by `arg0:int64,
+arg1: list int64, ret:int64`. Alternatively, if we don't care about
+the inputs after the call we can simplify it to `arg0:true, arg1:
+true, ret:int64`, where `true` is a predicate the is always satisfied.
+ 
+Then, using the linear implication `-o` we can express the
+specification of `is_elem` as
+
+```
+arg0:int64, arg1: list int64 -o arg0:true, arg1:true, ret:int64
+```
+
+Notice that this is a memory-safety specification, not a correctness
+predicate. We can reason about the correctness of the function once we
+extract it as a functional program in Coq.
+
+In the next sections, we shall define the predicates necessary to
+write this specification in Heapster.
+
+##### Defining permission predicates
+
+One of the simplest Heapster commands is `heapster_define_perm`, which defines
+a new named permission which can then be used in Heapster types. As an
+example, the following defines the permission which describes a 64-bit integer
+value:
+
+```
+heapster_define_perm env "int64" " " "llvmptr 64" "exists x:bv 64.eq(llvmword(x))"
+```
+
+The first argument is the Heapster environment, the second is its
+name, the third is its arguments (of which there are none), the fourth
+is the type of value that the permission applies to, and the fifth is
+its definition. the new permission is created and added to the
+environment.
```

Comment:
Mention that uses of named permissions are written like `int64<>`, where the `<>` is the empty list of arguments.